packages/core/content-manager/admin/src/pages/EditView/components/FormInputs/BlocksInput/BlocksToolbar.tsx

```diff
@@ -496,23 +524,170 @@ const LinkButton = ({ disabled }: { disabled: boolean }) => {
-        defaultMessage: 'Link',
-      }}
+      label={label}
       isActive={isLinkActive()}
       handleClick={addLink}
       disabled={isLinkDisabled()}
     />
   );
 };
 
+interface ObservedToolbarComponentProps {
+  index: number;
+  lastVisibleIndex: number;
+  setLastVisibleIndex: React.Dispatch<React.SetStateAction<number>>;
+  rootRef: React.RefObject<HTMLElement>;
+  children: React.ReactNode;
+}
+
+const ObservedToolbarComponent = ({
+  index,
+  lastVisibleIndex,
+  setLastVisibleIndex,
+  rootRef,
+  children,
+}: ObservedToolbarComponentProps) => {
+  const isVisible = index <= lastVisibleIndex;
+
+  const containerRef = useElementOnScreen<HTMLDivElement>(
+    (isVisible) => {
+      /**
+       * It's the MoreMenu's job to make an item not visible when there's not room for it.
+       * But we need to react here to the element becoming visible again.
+       */
+      if (isVisible) {
+        setLastVisibleIndex((prev) => Math.max(prev, index));
+      }
+    },
+    { threshold: 1, root: rootRef.current }
+  );
+
+  return (
+    <div
+      ref={containerRef}
+      style={{
+        /**
+         * Use visibility so that the element occupies the space if requires even when there's not
+         * enough room for it to be visible. The empty reserved space will be clipped by the
+         * overflow:hidden rule on the parent, so it doesn't affect the UI.
+         * This way we can keep observing its visiblity and react to browser resize events.
+         */
+        visibility: isVisible ? 'visible' : 'hidden',
+      }}
+    >
+      {children}
+    </div>
+  );
+};
+
+interface ObservedComponent {
+  toolbar: React.ReactNode;
+  menu: React.ReactNode;
+  key: string;
+}
+
+interface MoreMenuProps {
+  setLastVisibleIndex: React.Dispatch<React.SetStateAction<number>>;
+  hasHiddenItems: boolean;
+  rootRef: React.RefObject<HTMLElement>;
+  children: React.ReactNode;
+}
+
+const MoreMenu = ({ setLastVisibleIndex, hasHiddenItems, rootRef, children }: MoreMenuProps) => {
+  const { formatMessage } = useIntl();
+  const containerRef = useElementOnScreen<HTMLButtonElement>(
+    (isVisible) => {
+      // We only react to the menu becoming invisible. When that happens, we hide the last item.
+      if (!isVisible) {
+        /**
+         * If there's no room for any item, the index can be -1.
+         * This is intentional, in that case only the more menu will be visible.
+         **/
+        setLastVisibleIndex((prev) => prev - 1);
+      }
+    },
+    { threshold: 1, root: rootRef.current }
+  );
+
+  return (
+    <Menu.Root defaultOpen={false}>
+      <Menu.Trigger
+        endIcon={null}
+        paddingLeft={0}
+        paddingRight={0}
+        ref={containerRef}
+        style={{ visibility: hasHiddenItems ? 'visible' : 'hidden' }}
+      >
+        <IconButton
+          variant="ghost"
+          label={formatMessage({ id: 'global.more', defaultMessage: 'More' })}
+          tag="span"
+        >
```

Comment:
I think we will need the disabled state, and unfortunately I think you have to add it to both Menu.Trigger _and_ IconButotn